cpmpy/expressions/globalconstraints.py

```diff
@@ -274,18 +274,8 @@ def decompose_negation(self):
         should return something in negated normal form, since flatten_model.negated_normal() returns this
         '''
         from .python_builtins import all
-        succ = cpm_array(self.args)
-        n = len(succ)
-        order = intvar(0, n - 1, shape=n)
-        return [~all([AllDifferent(succ),
-                # different orders
-                AllDifferent(order)
-                    ]),
-                # not negating following constraints since they involve the auxiliary variables
-                # loop: first one is successor of '0'
-                order[0] == succ[0]
-                ] + [order[i] == succ[order[i - 1]] for i in range(1, n)]
-
+        decomp = self.decompose()
+        return [~all(decomp[:2])] + decomp[3:]
```

Comment:
this is cryptic and brittle...

cryptic: what does it do? why is it true?
how can others learn from this code?

brittle: what if somebody changes decomposition, or the order of the constraints in decompose...

Reply:
Hmm, yes this should probably be undone... 
I thought is was a good idea but maybe too cryptic indeed.
The main reasoning was: if we change the decomposition, we don't have to change it twice, but actually we do so yeah...